torch/_functorch/eager_transforms.py

```diff
@@ -975,26 +988,22 @@ def _jvp_with_argnums(func: Callable, primals: Any, tangents: Any, argnums: Opti
     diff_args = primals if argnums is None else _slice_argnums(primals, argnums)
     flat_primals, primals_spec = tree_flatten(diff_args)
     flat_tangents, tangents_spec = tree_flatten(tangents)
-    if primals_spec != tangents_spec:
-        raise RuntimeError(
-            f'{jvp_str}: Expected primals and tangents to have the same python '
-            f'structure. For example, if primals is a tuple of 3 tensors, '
-            f'tangents also must be. Got primals with structure {primals_spec} '
-            f'and tangents with structure {tangents_spec}')
+    _jvp_treespec_compare(diff_args, tangents)
     assert_non_empty_list_of_tensors(flat_primals, jvp_str, 'primals')
     assert_non_empty_list_of_tensors(flat_tangents, jvp_str, 'tangents')
 
-    level = _jvp_increment_nesting()
-    try:
-        global JVP_NESTING
-        JVP_NESTING += 1
+    with jvp_increment_nesting() as level:
         with fwAD._set_fwd_grad_enabled(True):
-            ctx = fwAD.dual_level if JVP_NESTING == 1 else noop
+            JVP_NESTING = torch._C._functorch.count_jvp_interpreters()
```

Comment:
I'm not sure that this is correct. JVP_NESTING is the number of nested jvps we're inside. `torch._C._functorch.count_jvp_interpreters()` counts the number of interpreters that are on the stack.

If we're in the middle of dispatching through the interpreters, the count of the interpreters != the number of jvps we're inside.

Reply:
Is it possible to have a jvp interpreter on the stack that was not created by calling `jvp_increment_nesting()`?

From the previous code, `JVP_NESTING` is incremented whenever `_jvp_increment_nesting()` is called (try block). So, if the former is not true, `JVP_NESTING` should reflect the number of times the context manager was executed, no?

Reply:
JVP_NESTING represents the number of nested context managers we're in.

The number of JVP interpreters that are on the stack is not the number of nested context managers that we're in. It is at most that number.

When we do a dispatch through a pytorch operation, the process of dispatching temporarily pops interpreters from the stack, so it's possible for the number of jvp interpreters to be less than the number of nested jvp context managers we're in.

Reply:
I've removed the `count_jvp_interpreters()` part and undo the changes related to it. Generated code should rely on the old `JVP_NESTING` variable. Could you review again once you have a chance?